lucene/core/src/java/org/apache/lucene/search/HitCountThresholdCheckerFactory.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Factory which generates instances of GlobalHitsThresholdChecker and LocalHitsThresholdChecker
+ * with the threshold as the passed in limit
+ */
+public final class HitCountThresholdCheckerFactory {
```

Comment:
Since we return a `HitsThresholdChecker` we could also make it public and add the private impl and the static functions creation there. We don't really need this factory class ?

Reply:
`HitsThresholdChecker` is currently an interface -- I am inclined to let it remain the same to allow users to do fancy early termination algorithms, hence introduced the factory. Thoughts?

Reply:
I agree for the interface but the static functions to help build simple checker could be directly implemented in `HitsThresholdChecker` ?

Reply:
> Since we return a `HitsThresholdChecker` we could also make it public and add the private impl and the static functions creation there. We don't really need this factory class ?

@jimczi  I was hesitant to do that since `HitsThresholdChecker` should really be agnostic of the underlying implementation. However, given the fact that the hit count checker(s) is the default, it makes sense to move it to the parent class. Updated, please take a look.